src/security/tls/tls_details.cpp

```diff
@@ -0,0 +1,530 @@
+
+    // prefix for extension's message signature
+    constexpr std::string_view sign_prefix = "libp2p-tls-handshake:";
+
+    // ASN1 tags
+    constexpr uint8_t kSequenceTag = 0x30;
+    constexpr uint8_t kOctetStringTag = 0x04;
+
+    // Size constants of the extension's format
+    constexpr size_t kMarshalledPublicKeySize = 36;
+    constexpr size_t kSignatureSize = 64;
+    constexpr size_t kAsnHeaderSize = 2;
+    constexpr size_t kSigOffset = 2 * kAsnHeaderSize + kMarshalledPublicKeySize;
+    constexpr size_t kExtensionDataSize =
+        kSigOffset + kAsnHeaderSize + kSignatureSize;
+
+    using Blob = std::vector<uint8_t>;
+    using Signature = std::array<uint8_t, kSignatureSize>;
+
+    // packs extension data into ASN1 sequence of 2 octet strings
+    std::array<uint8_t, kExtensionDataSize> marshalExtensionData(
+        const Blob &pub_key_bytes, const Signature &signature) {
+      if (pub_key_bytes.size() != kMarshalledPublicKeySize) {
+        throw std::runtime_error(
+            "unexpected size of peer's marshalled public key");
+      }
+
+      std::array<uint8_t, kExtensionDataSize> result{};
+
+      result[0] = kSequenceTag;
+
+      // data length
+      result[1] = kExtensionDataSize - kAsnHeaderSize;
+
+      result[2] = kOctetStringTag;
+      result[3] = kMarshalledPublicKeySize;
+      memcpy(&result[4], pub_key_bytes.data(), kMarshalledPublicKeySize);
+
+      result[kSigOffset] = kOctetStringTag;
+      result[kSigOffset + 1] = kSignatureSize;
+      memcpy(&result[kSigOffset + 2], signature.data(), kSignatureSize);
+
+      return result;
+    }
+
+    // make signature for the extension which ties certificate and host keys
+    Signature makeExtensionSignature(
+        const crypto::ecdsa::PublicKey &cert_pub_key,
+        const crypto::PrivateKey &host_private_key) {
+      constexpr size_t prefix_size = sign_prefix.size();
+      size_t msg_len = prefix_size + cert_pub_key.size();
+
+      uint8_t buf[msg_len];  // NOLINT
+      memcpy(buf, sign_prefix.data(), sign_prefix.size());
+      memcpy(buf + sign_prefix.size(),  // NOLINT
+             cert_pub_key.data(), cert_pub_key.size());
+
+      assert(host_private_key.type == crypto::Key::Type::Ed25519);
+      assert(host_private_key.data.size() == 32);
+
+      std::array<uint8_t, 32> pk_data{};
+      memcpy(pk_data.data(), host_private_key.data.data(), 32);
+
+      return crypto::ed25519::Ed25519ProviderImpl{}
+          .sign(gsl::span<const uint8_t>(buf, msg_len), pk_data)
+          .value();
+    }
+
+    // set certificate's public key
+    void assignPubkey(X509 *cert,
+                      const crypto::ecdsa::PublicKey &cert_pub_key) {
+      EVP_PKEY *pub = nullptr;
+      const auto *data = cert_pub_key.data();
+      d2i_PUBKEY(&pub, &data, cert_pub_key.size());
+      if (pub == nullptr) {
+        throw std::runtime_error("cannot deserialize certificate public key");
+      }
+      CLEANUP_PTR(pub, EVP_PKEY_free);
```

Comment:
```c++
auto free_pub = gsl::finally([pub]{EVP_PKEY_free(pub);});
```

Your approach will leak the memory in case of ```throw``` (e.g. on line 161).

Reply:
CLEANUP_PTR and gsl:: are effectively the same, no leak here